x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetMetaDataTestCase.java

```diff
@@ -64,6 +64,8 @@ public void testValidArrayTypes() throws IOException, SQLException {
                 typeName = typeName.equals("DATE") ? "DATETIME" : typeName;
                 typeName += "_ARRAY";
                 assertEquals(typeName, md.getColumnTypeName(i + 1));
+                // JdbcArray is not a public class
+                assertEquals("org.elasticsearch.xpack.sql.jdbc.JdbcArray", md.getColumnClassName(i + 1));
```

Comment:
the fact that it's a non public class is a give away that the return value is incorrect.

Reply:
I'm a bit hesitant here as I'm not sure if the docs aren't a bit conflicting: `Returns the fully-qualified name of the Java class whose instances are manufactured if the method ResultSet.getObject is called to retrieve a value from the column.`
So "org.elasticsearch.xpack.sql.jdbc.JdbcArray" is the one fitting this requirement.

But then also: `ResultSet.getObject may return a subclass of the class returned by this method.`
...so it might actually be a subclass and not those `whose instances are manufactured`.

Postgresql's driver obviously went for the superclass (from the above example).

Reply:
The issue here is with visibility; I believe this is better explained in the JDBC spec.
In short, as a consumer of the driver I'm using the methods here to perform discovery, for example:

`ResultSet.getObject(1, ClassLoader.loadClass(rsMetadata.getColumnClassName(1))`

or expect that
`ClassLoader.loadClass(rsMetadata.getColumnClassName()).isAssignable(resultSet.getObject(1).getClass())`

The above are problematic since either I need to use the driver classloader to load the private class or the object return won't match the classname (as it's not part of its hierarchy).
Essentially whatever the results are, they should be visible and useful to the user - anything else is a bug.

Reply:
That would make sense, but I'm not sure if having the implementation private and being able to load it by it's fully qualified class name are compatible?

Based on the docs (`ResultSet.getObject may return a subclass of the class returned by this method.`, i.e. `getColumnClassName()`), it would seem that returning the interface name for the class name might be the solution.

The JDBC spec doesn't cover this detail.
From what I can tell, some implementations make use of the ARRAY interface  - like [Solr's Calcite](https://gitbox.apache.org/repos/asf?p=solr.git;a=blob;f=solr/core/src/java/org/apache/solr/handler/CalciteJDBCStream.java;h=fd76cbf8f70acce35447c51119efce6ff258aa66;hb=HEAD#l51) and [Presto JDBC](https://github.com/prestodb/presto/blob/6ff3052e9df1a04613157746e7e779fd54c0c2a1/presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoResultSetMetaData.java#L230) - while others seem to use the implementation - like [Teradata](https://teradata-docs.s3.amazonaws.com/doc/connectivity/jdbc/reference/current/frameset.html) and [MySQL](https://github.com/mariadb-corporation/mariadb-connector-j/blob/d279517014f6cdcc86ef90154a5dd527c6e5a4ce/src/test/java/org/mariadb/jdbc/DatatypeTest.java#L235).

Reply:
> That would make sense, but I'm not sure if having the implementation private and being able to load it by it's fully qualified class name are compatible?

There's no value in returning a class that is private since one cannot use the class object directly (MyPrivate.class). Return a public parent (interface or otherwise) instead.

Reply:
Updated.